### DISCUSSION TITLE
How to make custom pages within Twill?

### DISCUSSION THREAD

**Discussion by adam-jones-net** [area17/twill#1542] [Q&A] [answered]
I need a few custom pages within my Twill admin that allow me to write whatever html/php that I want.  In other words, pages that use the twill page header and footer and detect that I've logged in to twill in order to access them.    Yet everything in between the header/footer is my own content.

Does anyone have code at hand to do this ?   There was something about this on the old spectrum chart forum but I can't find it now.

**Comment by ifox** [comment]
Hi @arkid, does this answer your needs: https://twill.io/docs/custom-cms-pages/ ?

**Reply by adam-jones-net** [reply]
Thanks @ifox I didn't realise it was in the docs these days. I know it wasn't years back when I first requested help on spectrum. Sorry for not checking first!

The official solution basically works fine however I seem to have found one issue with it. If I load the url of my custom page when not logged into Twill, instead of redirecting to the login page it throws an error.

I get a "Route [login] not defined" in the main Laravel route() function. vendor\laravel\framework\src\Illuminate\Routing\UrlGenerator.php

**Reply by ifox** [reply]
No worries. The free layout and it's docs have been merged in November 2019 and released in 1.2.3.

Which version of Twill are you on when encountering this error?

**Reply by adam-jones-net** [reply]
I can confirm that its 2.7.0

**Reply by ifox** [reply]
Thank you. Did you customize anything regarding Twill's exception handler?

**Reply by adam-jones-net** [reply]
I don't believe so but its a long time ago that I made the site.     Anywhere obvious I should check ?    Just to clarify; the redirecting to login works fine for all other twill page urls I've tried.

**Reply by adam-jones-net** [reply]
@ifox I've just made a totally fresh Laravel/Twill install (Laravel 8/Twill 2.7).   I created a custom page by the book as per the documentation and it's exactly the same.   So this is either classifiable as a bug or maybe some detail that is missing from the instructions for creating custom pages so that this gets handled properly?

![image](https://user-images.githubusercontent.com/3810108/160159347-a614c52c-dd92-413c-bb15-95e56245a5a2.png)

**Reply by ifox** [reply]
@arkid thanks for checking further. We're going to look into this. In the meantime, you can check the discussion on this PR to understand why this is happening: https://github.com/area17/twill/pull/1449.

**Comment by pboivin** [comment]
Hi @arkid, just a random thought, can you verify that your controller extends `A17\Twill\Http\Controllers\Admin\Controller`? I think the example in the docs uses a controller that doesn't extend anything. This might be the problem.

**Comment by adam-jones-net** [comment, marked as answer]
Bingo thanks @ifox that was it! :)     Can this be updated in the manual ?

What I now have is...
```
<?php
namespace App\Http\Controllers\Admin;
use A17\Twill\Http\Controllers\Admin\Controller as BaseController;

class CustomPageController extends BaseController {
    public function show()
    {
        return view('admin.customPage');
    }
}
```
